[Automation] java run fails for org.typelevel:cats-effect_3:3.1.1

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25454136977/job/74679302235
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
k ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 9s
11 actionable tasks: 1 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.typelevel:cats-effect_3:2.5.1 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 11s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 3.1.1 at 2026-05-06T18:50:47Z with exit code 1 after 11s
❌ ERROR: 'native-image run' for 3.1.1 failed with exit code 1.
SOFT-FAIL[native-image run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean test -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage javac compile
========================================================================================
Starting stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:47Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean compileTestJava -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 761ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 791ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:50Z with exit code 0 after 3s
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage java run
========================================================================================
Starting stage 'java run' for 3.1.1 at 2026-05-06T18:50:50Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 752ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:9:20: value Blocker is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:10:20: value ConcurrentEffect is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:11:20: value ContextShift is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:12:20: value ExitCase is not a member of cats.effect - did you mean effect.ExitCode?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:16:20: value Timer is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:17:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:18:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:19:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:20:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:38: Not found: type ContextShift
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:32: Not found: type ConcurrentEffect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:31: Not found: type Timer - did you mean Timeout?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:60: value contextShift is not a member of object cats.effect.IO
exception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:58: value ioConcurrentEffect is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:46: value timer is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:39:33: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.unsafe.IORuntime of parameter runtime of [33mmethod[0m [35munsafeRunTimed[0m in [33mclass[0m [35mIOPlatform[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:54:57: Found:    Unit
Required: cats.effect.IO[Option[cats.effect.IO[Unit]]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:75:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.Functor[cats.effect.IO] of parameter F of [33mmethod[0m [35mmake[0m in [33mobject[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:82:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.kernel.MonadCancel[cats.effect.IO, Throwable] of parameter F of [33mmethod[0m [35muse[0m in [33mclass[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:89:46: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:162:21: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:173:17: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:238:38: Not found: Blocker
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:265:64: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
24 errors found

> Task :compileTestScala FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
12 actionable tasks: 2 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 8s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 10s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.1.1 at 2026-05-06T18:51:01Z with exit code 1 after 11s
❌ ERROR: 'java run' for 3.1.1 failed with exit code 1.
FAILED[java run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25454136977/job/74679302165
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
e 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
12 actionable tasks: 2 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 9s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.typelevel:cats-effect_3:2.5.1 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 11s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 3.1.1 at 2026-05-06T18:50:36Z with exit code 1 after 12s
❌ ERROR: 'native-image run' for 3.1.1 failed with exit code 1.
SOFT-FAIL[native-image run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean test -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage javac compile
========================================================================================
Starting stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:36Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean compileTestJava -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 727ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 749ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:38Z with exit code 0 after 2s
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage java run
========================================================================================
Starting stage 'java run' for 3.1.1 at 2026-05-06T18:50:38Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 795ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava NO-SOURCE

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:9:20: value Blocker is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:10:20: value ConcurrentEffect is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:11:20: value ContextShift is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:12:20: value ExitCase is not a member of cats.effect - did you mean effect.ExitCode?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:16:20: value Timer is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:17:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:18:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:19:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:20:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:38: Not found: type ContextShift
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:32: Not found: type ConcurrentEffect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:31: Not found: type Timer - did you mean Timeout?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:60: value contextShift is not a member of object cats.effect.IOexception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner

[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:58: value ioConcurrentEffect is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:46: value timer is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:39:33: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.unsafe.IORuntime of parameter runtime of [33mmethod[0m [35munsafeRunTimed[0m in [33mclass[0m [35mIOPlatform[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:54:57: Found:    Unit
Required: cats.effect.IO[Option[cats.effect.IO[Unit]]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:75:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.Functor[cats.effect.IO] of parameter F of [33mmethod[0m [35mmake[0m in [33mobject[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:82:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.kernel.MonadCancel[cats.effect.IO, Throwable] of parameter F of [33mmethod[0m [35muse[0m in [33mclass[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:89:46: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:162:21: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:173:17: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:238:38: Not found: Blocker
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:265:64: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
24 errors found

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

> Task :compileTestScala FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

* Try:
Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
12 actionable tasks: 2 executed, 10 up-to-date

> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 8s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 10s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.1.1 at 2026-05-06T18:50:49Z with exit code 1 after 11s
❌ ERROR: 'java run' for 3.1.1 failed with exit code 1.
FAILED[java run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25454136977/job/74679302249
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
were used in this build, making it incompatible with Gradle 10.
Execution failed for task ':compileTestScala'.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
11 actionable tasks: 1 executed, 10 up-to-date
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 9s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.typelevel:cats-effect_3:2.5.1 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 10s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 3.1.1 at 2026-05-06T18:50:54Z with exit code 1 after 11s
❌ ERROR: 'native-image run' for 3.1.1 failed with exit code 1.
SOFT-FAIL[native-image run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.1.1" ./gradlew clean test -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage javac compile
========================================================================================
Starting stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:54Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.1.1" ./gradlew clean compileTestJava -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 721ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 744ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.1.1 at 2026-05-06T18:50:56Z with exit code 0 after 2s
========================================================================================
 org.typelevel:cats-effect_3:2.5.1:3.1.1 stage java run
========================================================================================
Starting stage 'java run' for 3.1.1 at 2026-05-06T18:50:56Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 814ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:9:20: value Blocker is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:10:20: value ConcurrentEffect is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:11:20: value ContextShift is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:12:20: value ExitCase is not a member of cats.effect - did you mean effect.ExitCode?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:16:20: value Timer is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:17:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:18:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:19:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:20:20: value concurrent is not a member of cats.effect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:38: Not found: type ContextShift
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:32: Not found: type ConcurrentEffect
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:31: Not found: type Timer - did you mean Timeout?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:34:60: value contextShift is not a member of object cats.effect.IO
exception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:35:58: value ioConcurrentEffect is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:36:46: value timer is not a member of object cats.effect.IO
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:39:33: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.unsafe.IORuntime of parameter runtime of [33mmethod[0m [35munsafeRunTimed[0m in [33mclass[0m [35mIOPlatform[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:54:57: Found:    Unit
Required: cats.effect.IO[Option[cats.effect.IO[Unit]]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:75:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.Functor[cats.effect.IO] of parameter F of [33mmethod[0m [35mmake[0m in [33mobject[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:82:9: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type cats.effect.kernel.MonadCancel[cats.effect.IO, Throwable] of parameter F of [33mmethod[0m [35muse[0m in [33mclass[0m [35mResource[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:89:46: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:162:21: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:173:17: Not found: type ExitCase
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:238:38: Not found: Blocker
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala:265:64: Ambiguous given instances: both [33mvalue[0m [35mcontextShift[0m in [33mclass[0m [35mCats_effect_3Test[0m and [33mvalue[0m [35meffect[0m in [33mclass[0m [35mCats_effect_3Test[0m match type Null <:< A1 of parameter ev of [33mmethod[0m [35morNull[0m in [33mclass[0m [35mOption[0m
24 errors found

FAILURE: Build failed with an exception.

> Task :compileTestScala FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.typelevel/cats-effect_3/2.5.1/build/reports/problems/problems-report.html

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED12 actionable tasks: 2 executed, 10 up-to-date
 in 8s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 10s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.1.1 at 2026-05-06T18:51:07Z with exit code 1 after 11s
❌ ERROR: 'java run' for 3.1.1 failed with exit code 1.
FAILED[java run][3.1.1][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.1.1" ./gradlew clean javaTest -Pcoordinates="org.typelevel:cats-effect_3:2.5.1"]
```

--- a/metadata/org.typelevel/cats-effect_3/3.1.1/reachability-metadata.json
+++ b/metadata/org.typelevel/cats-effect_3/3.1.1/reachability-metadata.json
@@ -1,0 +1,37 @@
+{
+  "reflection" : [
+    {
+      "condition" : {
+        "typeReached" : "cats.effect.std.Semaphore$impl"
+      },
+      "type" : "cats.effect.std.Semaphore$impl",
+      "fields" : [
+        {
+          "name" : "0bitmap$1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "cats.effect.unsafe.IORuntimeCompanionPlatform"
+      },
+      "type" : "cats.effect.unsafe.IORuntimeCompanionPlatform",
+      "fields" : [
+        {
+          "name" : "0bitmap$1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "cats.effect.unsafe.WorkStealingThreadPool"
+      },
+      "type" : "cats.effect.unsafe.WorkStealingThreadPool",
+      "fields" : [
+        {
+          "name" : "0bitmap$1"
+        }
+      ]
+    }
+  ]
+}

--- a/metadata/org.typelevel/cats-effect_3/index.json
+++ b/metadata/org.typelevel/cats-effect_3/index.json
@@ -1,6 +1,19 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "3.1.1",
+    "language" : {
+      "name" : "scala",
+      "version" : "3"
+    },
+    "tested-versions" : [
+      "3.1.1"
+    ],
+    "allowed-packages" : [
+      "cats.effect"
+    ]
+  },
+  {
     "metadata-version" : "2.5.1",
     "source-code-url" : "https://repo1.maven.org/maven2/org/typelevel/cats-effect_3/2.5.1/cats-effect_3-2.5.1-sources.jar",
     "repository-url" : "https://github.com/typelevel/cats-effect",

--- a/metadata/org.typelevel/cats-effect_3/index.json
+++ b/metadata/org.typelevel/cats-effect_3/index.json
@@ -2,6 +2,11 @@
   {
     "latest" : true,
     "metadata-version" : "3.1.1",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/org/typelevel/cats-effect_3/$version$/cats-effect_3-$version$-sources.jar",
+    "repository-url" : "https://github.com/typelevel/cats-effect",
+    "test-code-url" : "https://github.com/typelevel/cats-effect/tree/v$version$/tests",
+    "documentation-url" : "https://github.com/typelevel/cats-effect/tree/v$version$/docs",
+    "description" : "Cats Effect is a high-performance, asynchronous runtime and effect system for building purely functional Scala applications. It provides the IO data type, resource-safety tools, concurrency primitives, and type classes for describing and controlling side effects.",
     "language" : {
       "name" : "scala",
       "version" : "3"

--- a/stats/org.typelevel/cats-effect_3/3.1.1/execution-metrics.json
+++ b/stats/org.typelevel/cats-effect_3/3.1.1/execution-metrics.json
@@ -1,0 +1,87 @@
+{
+  "fix_java_run_fail:2026-05-06": {
+    "timestamp": "2026-05-06T22:36:45.845935Z",
+    "library": "org.typelevel:cats-effect_3:3.1.1",
+    "starting_commit": "5d3b993e77a0997873697b8f0e3c2e3392d6efc8",
+    "ending_commit": "965ec52ca557c0a02fec4ba4bb6b791bbeb214bf",
+    "previous_library": "org.typelevel:cats-effect_3:2.5.1",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "3.1.1",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 5333,
+          "missed": 10569,
+          "ratio": 0.335367,
+          "total": 15902
+        },
+        "line": {
+          "covered": 1062,
+          "missed": 823,
+          "ratio": 0.563395,
+          "total": 1885
+        },
+        "method": {
+          "covered": 436,
+          "missed": 1114,
+          "ratio": 0.28129,
+          "total": 1550
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "2.5.1",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 5802,
+          "missed": 23596,
+          "ratio": 0.19736,
+          "total": 29398
+        },
+        "line": {
+          "covered": 962,
+          "missed": 1908,
+          "ratio": 0.335192,
+          "total": 2870
+        },
+        "method": {
+          "covered": 567,
+          "missed": 2644,
+          "ratio": 0.176581,
+          "total": 3211
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 76682,
+      "cached_input_tokens_used": 851456,
+      "output_tokens_used": 9790,
+      "iterations": 3,
+      "cost_usd": 0.7194,
+      "tested_library_loc": 1062,
+      "metadata_entries": 6,
+      "previous_library_metadata_entries": 10,
+      "code_coverage_percent": 56.34,
+      "previous_library_coverage_percent": 33.52
+    },
+    "artifacts": {
+      "test_file": "tests/src/org.typelevel/cats-effect_3/3.1.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala",
+      "metadata_file": "metadata/org.typelevel/cats-effect_3/3.1.1/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/org.typelevel/cats-effect_3/3.1.1/stats.json
+++ b/stats/org.typelevel/cats-effect_3/3.1.1/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "3.1.1",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 5333,
+        "missed" : 10569,
+        "ratio" : 0.335367,
+        "total" : 15902
+      },
+      "line" : {
+        "covered" : 1062,
+        "missed" : 823,
+        "ratio" : 0.563395,
+        "total" : 1885
+      },
+      "method" : {
+        "covered" : 436,
+        "missed" : 1114,
+        "ratio" : 0.28129,
+        "total" : 1550
+      }
+    }
+  } ]
+}

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/.gitignore
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/build.gradle
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/build.gradle
@@ -1,0 +1,36 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+    id "scala"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.typelevel:cats-effect_3:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+    testImplementation 'org.scala-lang:scala3-library_3:3.3.6'
+    testImplementation 'org.scala-lang:scala3-compiler_3:3.3.6'
+}
+
+java {
+    toolchain {
+        languageVersion = JavaLanguageVersion.of(21)
+    }
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/gradle.properties
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = org.typelevel:cats-effect_3:3.1.1
+library.version = 3.1.1
+metadata.dir = org.typelevel/cats-effect_3/3.1.1/

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/settings.gradle
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'org.typelevel.cats-effect_3_tests'

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala
@@ -1,0 +1,267 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_typelevel.cats_effect_3
+
+import cats.effect.Blocker
+import cats.effect.ConcurrentEffect
+import cats.effect.ContextShift
+import cats.effect.ExitCase
+import cats.effect.IO
+import cats.effect.Resource
+import cats.effect.SyncIO
+import cats.effect.Timer
+import cats.effect.concurrent.Deferred
+import cats.effect.concurrent.MVar
+import cats.effect.concurrent.Ref
+import cats.effect.concurrent.Semaphore
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.Timeout
+
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+class Cats_effect_3Test {
+  private implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
+  private implicit val effect: ConcurrentEffect[IO] = IO.ioConcurrentEffect(contextShift)
+  private implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
+
+  private def await[A](io: IO[A]): A = {
+    io.unsafeRunTimed(5.seconds).getOrElse(throw new AssertionError("IO did not complete within the expected timeout"))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def ioComposesLazySynchronousAndAsynchronousEffects(): Unit = {
+    var evaluations: Int = 0
+    val failure: IllegalStateException = new IllegalStateException("expected failure")
+
+    val program: IO[(Int, String, Int)] = for {
+      delayed <- IO {
+        evaluations += 1
+        40
+      }.map(_ + 2)
+      recovered <- IO.raiseError[Int](failure).attempt.map(_.fold(_.getMessage, _.toString))
+      asynchronous <- IO.async[Int](callback => callback(Right(21))).map(_ * 2)
+    } yield (delayed, recovered, asynchronous)
+
+    val result: (Int, String, Int) = await(program)
+
+    assertThat(evaluations).isEqualTo(1)
+    assertThat(result).isEqualTo((42, "expected failure", 42))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def resourceRunsFinalizersForSuccessfulAndFailedUses(): Unit = {
+    val events: ListBuffer[String] = ListBuffer.empty[String]
+    val failure: RuntimeException = new RuntimeException("use failed")
+
+    def managed(label: String): Resource[IO, String] = {
+      Resource.make(IO {
+        events += s"acquire-$label"
+        label
+      })(value => IO {
+        events += s"release-$value"
+      })
+    }
+
+    val program: IO[(Int, Either[Throwable, Int])] = for {
+      success <- managed("success").use(value => IO {
+        events += s"use-$value"
+        value.length
+      })
+      failed <- managed("failure").use(_ => IO.raiseError[Int](failure)).attempt
+    } yield (success, failed)
+
+    val result: (Int, Either[Throwable, Int]) = await(program)
+
+    assertThat(result._1).isEqualTo(7)
+    assertThat(result._2.left.toOption.orNull).isSameAs(failure)
+    assertThat(events.toList).isEqualTo(List(
+      "acquire-success",
+      "use-success",
+      "release-success",
+      "acquire-failure",
+      "release-failure"
+    ))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def refSupportsAtomicUpdatesModificationAndCompareAndSetAccess(): Unit = {
+    val program: IO[(Int, Boolean, Map[String, Int])] = for {
+      ref <- Ref.of[IO, Map[String, Int]](Map.empty[String, Int])
+      _ <- ref.update(_.updated("alpha", 1))
+      previous <- ref.modify { values =>
+        val nextValue: Int = values("alpha") + 1
+        (values.updated("alpha", nextValue), values("alpha"))
+      }
+      accessResult <- ref.access.flatMap { case (snapshot, setter) =>
+        setter(snapshot.updated("beta", 3)).flatMap(success => ref.get.map(values => (success, values)))
+      }
+    } yield (previous, accessResult._1, accessResult._2)
+
+    val result: (Int, Boolean, Map[String, Int]) = await(program)
+
+    assertThat(result).isEqualTo((1, true, Map("alpha" -> 2, "beta" -> 3)))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def deferredCoordinatesFibersWithoutBlockingThreads(): Unit = {
+    val program: IO[(String, Vector[String])] = for {
+      gate <- Deferred[IO, String]
+      observed <- Ref.of[IO, Vector[String]](Vector.empty[String])
+      fiber <- gate.get.flatMap(value => observed.update(_ :+ value).map(_ => value.reverse)).start
+      beforeComplete <- observed.get
+      _ <- gate.complete("ready")
+      joined <- fiber.join
+      afterComplete <- observed.get
+    } yield (s"$beforeComplete:$joined", afterComplete)
+
+    val result: (String, Vector[String]) = await(program)
+
+    assertThat(result).isEqualTo(("Vector():ydaer", Vector("ready")))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def semaphoreAndMVarCoordinatePermitProtectedExchange(): Unit = {
+    val program: IO[(Boolean, Boolean, String, Long)] = for {
+      semaphore <- Semaphore[IO](1L)
+      mailbox <- MVar[IO].empty[String]
+      _ <- semaphore.acquire
+      acquireWhileHeld <- semaphore.tryAcquire
+      _ <- semaphore.release
+      acquireAfterRelease <- semaphore.tryAcquire
+      _ <- if (acquireAfterRelease) semaphore.release else IO.unit
+      producer <- semaphore.withPermit(mailbox.put("payload")).start
+      received <- mailbox.take
+      _ <- producer.join
+      permits <- semaphore.available
+    } yield (acquireWhileHeld, acquireAfterRelease, received, permits)
+
+    val result: (Boolean, Boolean, String, Long) = await(program)
+
+    assertThat(result).isEqualTo((false, true, "payload", 1L))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def cancelingFiberRunsBracketCaseFinalizerWithCanceledExitCase(): Unit = {
+    val program: IO[ExitCase[Throwable]] = for {
+      started <- Deferred[IO, Unit]
+      canceled <- Deferred[IO, ExitCase[Throwable]]
+      fiber <- started.complete(()).bracketCase(_ => IO.never) { (_, exitCase) =>
+        canceled.complete(exitCase)
+      }.start
+      _ <- started.get
+      _ <- fiber.cancel
+      exitCase <- canceled.get.timeoutTo(1.second, IO.raiseError(new AssertionError("cancellation finalizer did not run")))
+    } yield exitCase
+
+    val result: ExitCase[Throwable] = await(program)
+
+    assertThat(result).isEqualTo(ExitCase.Canceled)
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def raceReturnsTheFirstCompletedEffect(): Unit = {
+    val program: IO[Either[String, String]] = IO.race(
+      timer.sleep(1.second).map(_ => "slow"),
+      IO.pure("fast")
+    )
+
+    val result: Either[String, String] = await(program)
+
+    assertThat(result).isEqualTo(Right("fast"))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def timerAndTimeoutToProduceBoundedFallbacks(): Unit = {
+    val neverCompletes: IO[Int] = IO.never
+    val program: IO[(Int, Boolean)] = for {
+      before <- timer.clock.monotonic(TimeUnit.MILLISECONDS)
+      value <- neverCompletes.timeoutTo(25.millis, IO.pure(99))
+      after <- timer.clock.monotonic(TimeUnit.MILLISECONDS)
+    } yield (value, after >= before)
+
+    val result: (Int, Boolean) = await(program)
+
+    assertThat(result).isEqualTo((99, true))
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def contextShiftEvaluatesEffectsOnSpecifiedExecutionContext(): Unit = {
+    val executor: ExecutorService = Executors.newSingleThreadExecutor(new ThreadFactory {
+      override def newThread(runnable: Runnable): Thread = {
+        val thread: Thread = new Thread(runnable)
+        thread.setDaemon(true)
+        thread.setName("cats-effect-eval-on-test")
+        thread
+      }
+    })
+    val dedicatedExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(executor)
+
+    try {
+      val program: IO[(String, String)] = for {
+        shiftedThreadName <- contextShift.evalOn(dedicatedExecutionContext)(IO(Thread.currentThread().getName))
+        resumedThreadName <- IO(Thread.currentThread().getName)
+      } yield (shiftedThreadName, resumedThreadName)
+
+      val result: (String, String) = await(program)
+
+      assertThat(result._1).startsWith("cats-effect-eval-on-test")
+      assertThat(result._2).doesNotStartWith("cats-effect-eval-on-test")
+    } finally {
+      executor.shutdownNow()
+      executor.awaitTermination(1, TimeUnit.SECONDS)
+    }
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def blockerEvaluatesBlockingWorkInsideManagedResource(): Unit = {
+    val program: IO[(String, Int)] = Blocker[IO].use { blocker =>
+      for {
+        threadName <- blocker.blockOn(IO(Thread.currentThread().getName))
+        computed <- blocker.delay(21 + 21)
+      } yield (threadName, computed)
+    }
+
+    val result: (String, Int) = await(program)
+
+    assertThat(result._1).isNotBlank
+    assertThat(result._2).isEqualTo(42)
+  }
+
+  @Test
+  @Timeout(value = 10, unit = TimeUnit.SECONDS)
+  def syncIOEvaluatesPureSynchronousEffectsAndCapturesFailures(): Unit = {
+    var counter: Int = 0
+
+    val result: (Int, Either[Throwable, Int]) = SyncIO {
+      counter += 1
+      counter
+    }.flatMap(value => SyncIO.pure(value + 1))
+      .flatMap(value => SyncIO.raiseError[Int](new IllegalArgumentException(value.toString)).attempt.map(error => (value, error)))
+      .unsafeRunSync()
+
+    assertThat(counter).isEqualTo(1)
+    assertThat(result._1).isEqualTo(2)
+    assertThat(result._2.left.toOption.map(_.getMessage).orNull).isEqualTo("2")
+  }
+}

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/src/test/scala/org_typelevel/cats_effect_3/Cats_effect_3Test.scala
@@ -6,18 +6,15 @@
  */
 package org_typelevel.cats_effect_3
 
-import cats.effect.Blocker
-import cats.effect.ConcurrentEffect
-import cats.effect.ContextShift
-import cats.effect.ExitCase
+import cats.effect.Deferred
 import cats.effect.IO
+import cats.effect.OutcomeIO
+import cats.effect.Ref
 import cats.effect.Resource
 import cats.effect.SyncIO
-import cats.effect.Timer
-import cats.effect.concurrent.Deferred
-import cats.effect.concurrent.MVar
-import cats.effect.concurrent.Ref
-import cats.effect.concurrent.Semaphore
+import cats.effect.std.Queue
+import cats.effect.std.Semaphore
+import cats.effect.unsafe.implicits.global
 import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.Timeout
@@ -31,10 +28,6 @@ import scala.concurrent.ExecutionContext
 import scala.concurrent.duration._
 
 class Cats_effect_3Test {
-  private implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
-  private implicit val effect: ConcurrentEffect[IO] = IO.ioConcurrentEffect(contextShift)
-  private implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
-
   private def await[A](io: IO[A]): A = {
     io.unsafeRunTimed(5.seconds).getOrElse(throw new AssertionError("IO did not complete within the expected timeout"))
   }
@@ -51,7 +44,7 @@ class Cats_effect_3Test {
         40
       }.map(_ + 2)
       recovered <- IO.raiseError[Int](failure).attempt.map(_.fold(_.getMessage, _.toString))
-      asynchronous <- IO.async[Int](callback => callback(Right(21))).map(_ * 2)
+      asynchronous <- IO.async_[Int](callback => callback(Right(21))).map(_ * 2)
     } yield (delayed, recovered, asynchronous)
 
     val result: (Int, String, Int) = await(program)
@@ -125,7 +118,7 @@ class Cats_effect_3Test {
       fiber <- gate.get.flatMap(value => observed.update(_ :+ value).map(_ => value.reverse)).start
       beforeComplete <- observed.get
       _ <- gate.complete("ready")
-      joined <- fiber.join
+      joined <- fiber.joinWithNever
       afterComplete <- observed.get
     } yield (s"$beforeComplete:$joined", afterComplete)
 
@@ -136,18 +129,18 @@ class Cats_effect_3Test {
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
-  def semaphoreAndMVarCoordinatePermitProtectedExchange(): Unit = {
+  def semaphoreAndQueueCoordinatePermitProtectedExchange(): Unit = {
     val program: IO[(Boolean, Boolean, String, Long)] = for {
       semaphore <- Semaphore[IO](1L)
-      mailbox <- MVar[IO].empty[String]
+      mailbox <- Queue.bounded[IO, String](1)
       _ <- semaphore.acquire
       acquireWhileHeld <- semaphore.tryAcquire
       _ <- semaphore.release
       acquireAfterRelease <- semaphore.tryAcquire
       _ <- if (acquireAfterRelease) semaphore.release else IO.unit
-      producer <- semaphore.withPermit(mailbox.put("payload")).start
+      producer <- semaphore.permit.use(_ => mailbox.offer("payload")).start
       received <- mailbox.take
-      _ <- producer.join
+      _ <- producer.joinWithNever
       permits <- semaphore.available
     } yield (acquireWhileHeld, acquireAfterRelease, received, permits)
 
@@ -158,28 +151,28 @@ class Cats_effect_3Test {
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
-  def cancelingFiberRunsBracketCaseFinalizerWithCanceledExitCase(): Unit = {
-    val program: IO[ExitCase[Throwable]] = for {
+  def cancelingFiberRunsBracketCaseFinalizerWithCanceledOutcome(): Unit = {
+    val program: IO[OutcomeIO[Unit]] = for {
       started <- Deferred[IO, Unit]
-      canceled <- Deferred[IO, ExitCase[Throwable]]
-      fiber <- started.complete(()).bracketCase(_ => IO.never) { (_, exitCase) =>
-        canceled.complete(exitCase)
+      canceled <- Deferred[IO, OutcomeIO[Unit]]
+      fiber <- IO.unit.bracketCase(_ => started.complete(()).flatMap(_ => IO.never[Unit])) { (_, outcome) =>
+        canceled.complete(outcome).map(_ => ())
       }.start
       _ <- started.get
       _ <- fiber.cancel
-      exitCase <- canceled.get.timeoutTo(1.second, IO.raiseError(new AssertionError("cancellation finalizer did not run")))
-    } yield exitCase
+      outcome <- canceled.get.timeoutTo(1.second, IO.raiseError(new AssertionError("cancellation finalizer did not run")))
+    } yield outcome
 
-    val result: ExitCase[Throwable] = await(program)
+    val result: OutcomeIO[Unit] = await(program)
 
-    assertThat(result).isEqualTo(ExitCase.Canceled)
+    assertThat(result.fold("canceled", _ => "errored", _ => "succeeded")).isEqualTo("canceled")
   }
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
   def raceReturnsTheFirstCompletedEffect(): Unit = {
     val program: IO[Either[String, String]] = IO.race(
-      timer.sleep(1.second).map(_ => "slow"),
+      IO.sleep(1.second).map(_ => "slow"),
       IO.pure("fast")
     )
 
@@ -190,12 +183,12 @@ class Cats_effect_3Test {
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
-  def timerAndTimeoutToProduceBoundedFallbacks(): Unit = {
+  def monotonicClockAndTimeoutToProduceBoundedFallbacks(): Unit = {
     val neverCompletes: IO[Int] = IO.never
     val program: IO[(Int, Boolean)] = for {
-      before <- timer.clock.monotonic(TimeUnit.MILLISECONDS)
+      before <- IO.monotonic
       value <- neverCompletes.timeoutTo(25.millis, IO.pure(99))
-      after <- timer.clock.monotonic(TimeUnit.MILLISECONDS)
+      after <- IO.monotonic
     } yield (value, after >= before)
 
     val result: (Int, Boolean) = await(program)
@@ -205,7 +198,7 @@ class Cats_effect_3Test {
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
-  def contextShiftEvaluatesEffectsOnSpecifiedExecutionContext(): Unit = {
+  def evalOnEvaluatesEffectsOnSpecifiedExecutionContext(): Unit = {
     val executor: ExecutorService = Executors.newSingleThreadExecutor(new ThreadFactory {
       override def newThread(runnable: Runnable): Thread = {
         val thread: Thread = new Thread(runnable)
@@ -218,7 +211,7 @@ class Cats_effect_3Test {
 
     try {
       val program: IO[(String, String)] = for {
-        shiftedThreadName <- contextShift.evalOn(dedicatedExecutionContext)(IO(Thread.currentThread().getName))
+        shiftedThreadName <- IO(Thread.currentThread().getName).evalOn(dedicatedExecutionContext)
         resumedThreadName <- IO(Thread.currentThread().getName)
       } yield (shiftedThreadName, resumedThreadName)
 
@@ -234,13 +227,11 @@ class Cats_effect_3Test {
 
   @Test
   @Timeout(value = 10, unit = TimeUnit.SECONDS)
-  def blockerEvaluatesBlockingWorkInsideManagedResource(): Unit = {
-    val program: IO[(String, Int)] = Blocker[IO].use { blocker =>
-      for {
-        threadName <- blocker.blockOn(IO(Thread.currentThread().getName))
-        computed <- blocker.delay(21 + 21)
-      } yield (threadName, computed)
-    }
+  def ioBlockingEvaluatesBlockingWorkOnBlockingPool(): Unit = {
+    val program: IO[(String, Int)] = for {
+      threadName <- IO.blocking(Thread.currentThread().getName)
+      computed <- IO.blocking(21 + 21)
+    } yield (threadName, computed)
 
     val result: (String, Int) = await(program)
 

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/user-code-filter.json
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/user-code-filter.json
@@ -1,0 +1,10 @@
+{
+  "rules": [
+    {
+      "excludeClasses": "**"
+    },
+    {
+      "includeClasses": "cats.effect.**"
+    }
+  ]
+}

--- a/tests/src/org.typelevel/cats-effect_3/3.1.1/user-code-filter.json
+++ b/tests/src/org.typelevel/cats-effect_3/3.1.1/user-code-filter.json
@@ -1,10 +1,13 @@
 {
-  "rules": [
+  "rules" : [
     {
-      "excludeClasses": "**"
+      "excludeClasses" : "**"
     },
     {
-      "includeClasses": "cats.effect.**"
+      "includeClasses" : "cats.effect.**"
+    },
+    {
+      "includeClasses" : "org_typelevel.cats_effect_3.**"
     }
   ]
 }

